xapian-letor/tests/api_letor.cc

```diff
@@ -185,6 +454,139 @@ DEFINE_TESTCASE(listnet_ranker, generated)
+    string training_data = data_directory + "training_data_one_document.txt";
+    ranker.set_database_path(db_path);
+    TEST_EQUAL(ranker.get_database_path(), db_path);
+    ranker.set_query(Xapian::Query("tigers"));
+    ranker.train_model(training_data);
+    Xapian::docid doc1 = *mymset[0];
+    ranker.rank(mymset);
+    TEST_EQUAL(doc1, *mymset[0]);
+    mymset = enquire.get_mset(0, 10);
+    ranker.train_model(training_data, "ListNet_Ranker");
+    ranker.rank(mymset, "ListNet_Ranker");
+    TEST_EQUAL(doc1, *mymset[0]);
+    TEST_EXCEPTION(Xapian::LetorInternalError,
+		   ranker.score(query, qrel, "ListNet_Ranker",
+				"scorer_output1.txt", 10, ""));
+    TEST_EXCEPTION(Xapian::FileNotFoundError,
+		   ranker.score("", qrel, "ListNet_Ranker",
+				"scorer_output1.txt", 10));
+    TEST_EXCEPTION(Xapian::FileNotFoundError,
+		   ranker.score(qrel, "", "ListNet_Ranker",
+				"scorer_output1.txt", 10));
+    ranker.score(query, qrel, "ListNet_Ranker", "ndcg_output1.txt", 10);
+    ranker.score(query, qrel, "ListNet_Ranker", "err_output1.txt", 10, "ERRScore");
+
+    return true;
+}
+
+
+// DEFINE_TESTCASE(listnet_ranker_three_miss, generated)
+// {
+//     Xapian::ListNETRanker ranker;
+//     TEST_EXCEPTION(Xapian::FileNotFoundError, ranker.train_model(""));
+//     string db_path = get_database_path("apitest_listnet_ranker3",
+//					  db_index_three_documents);
+//     Xapian::Enquire enquire((Xapian::Database(db_path)));
+//     enquire.set_query(Xapian::Query("score"));
+//     Xapian::MSet mymset = enquire.get_mset(0, 10);
+//     string data_directory = test_driver::get_srcdir() + "/testdata/";
+//     string query = data_directory + "querythree.txt";
+//     string qrel = data_directory + "qrelthree_onemiss.txt";
+//     string training_data = data_directory + "training_data_three_onemiss.txt";
+//     ranker.set_database_path(db_path);
+//     TEST_EQUAL(ranker.get_database_path(), db_path);
+//     ranker.set_query(Xapian::Query("score"));
+//     ranker.train_model(training_data);
+//     TEST_EQUAL(mymset.size(), 2);
+//     Xapian::docid doc1 = *mymset[0];
+//     Xapian::docid doc2 = *mymset[1];
+//     ranker.rank(mymset);
+//     TEST_EQUAL(doc1, *mymset[1]);
+//     TEST_EQUAL(doc2, *mymset[0]);
+//     mymset = enquire.get_mset(0, 10);
+//     ranker.train_model(training_data, "ListNet_Ranker");
+//     ranker.rank(mymset, "ListNet_Ranker");
+//     TEST_EQUAL(doc1, *mymset[1]);
+//     TEST_EQUAL(doc2, *mymset[0]);
+//     TEST_EXCEPTION(Xapian::LetorInternalError,
+//			ranker.score(query, qrel, "ListNet_Ranker",
+//			"scorer_output2.txt", 10, ""));
+//     TEST_EXCEPTION(Xapian::FileNotFoundError,
+//				ranker.score("", qrel, "ListNet_Ranker",
+//				"scorer_output2.txt", 10));
+//     TEST_EXCEPTION(Xapian::FileNotFoundError,
+//			ranker.score(qrel, "", "ListNet_Ranker",
+//						"scorer_output.txt2", 10));
+//     ranker.score(query, qrel, "ListNet_Ranker", "ndcg_output2.txt", 10);
```

Comment:
https://pastebin.com/EnQTu8Ve
Would be the failure when any of the missing-qrel tests is uncommented.